File: preppy/manifest.py
"""Manifest builder: the viewer-native, per-object descriptor that replaces the
Voyager ``items.json`` + ``*.svx.json`` pair (ADR-0002, amended).

The ``<dri-viewer>`` widget consumes **one manifest per object** (one scene):
object-level metadata (``title``, ``inventory``, ``description``, ``credit``,
``date``, ``units``) plus a flat ``variants[]``, each a self-contained glb
referenced by ``uri`` with a stable ``id`` (the variant ``suffix``, used for
deep-linking) and a display ``label``. Exactly one variant is flagged
``default``. Variants may carry per-variant provenance overrides (``credit``,
``date``, ``method``, ``description``).

An optional flat ``index.json`` lists objects for a host archive UI; the widget
does not require it.

Builders are pure dict factories (no I/O) so they unit-test without the
toolchain; :func:`write_json` is the only side-effecting helper.
"""

import json
from pathlib import Path
from typing import Dict, List, Mapping, Optional, Sequence, Union

PathLike = Union[str, Path]

#: Default measurement units for a scene (spike: geometry is in cm).
DEFAULT_UNITS = 'cm'

#: Object-level metadata copied verbatim into the manifest, in manifest order.
#: ``id`` and ``units`` are handled explicitly; ``variants`` is appended last.
OBJECT_META_FIELDS = ('title', 'titles', 'inventory', 'description',
                      'credit', 'date')

#: Optional per-variant provenance overrides copied into a variant entry.
VARIANT_OVERRIDE_FIELDS = ('credit', 'date', 'method', 'description')


def default_variant_index(variants: Sequence[Mapping]) -> int:
    """Return the index of the variant to mark ``default``.

    The first variant whose config sets ``"default": true`` wins; if none do,
    the first variant is the default. Raises if ``variants`` is empty.
    """
    if not variants:
        raise ValueError('an object must declare at least one variant')
    for i, v in enumerate(variants):
        if v.get('default'):
            return i
    return 0


def variant_entry(variant_cfg: Mapping, uri: str, *,
                  default: bool = False) -> Dict:
    """Build one manifest ``variants[]`` entry.

    ``id`` is the variant's stable ``suffix`` (deep-link / selection key),
    ``label`` its display name (falls back to the suffix), ``uri`` the emitted
    self-contained glb. ``default`` adds ``"default": true`` only when set (the
    non-default entries omit the key). Present, non-null override fields
    (:data:`VARIANT_OVERRIDE_FIELDS`) are copied through.
    """
    try:
        vid = variant_cfg['suffix']
    except KeyError:
        raise KeyError("variant config requires a 'suffix'")

    entry: Dict = {
        'id': vid,
        'label': variant_cfg.get('label', vid),
        'uri': uri,
    }
    if default:
        entry['default'] = True
    for field in VARIANT_OVERRIDE_FIELDS:
        val = variant_cfg.get(field)
        if val is not None:
            entry[field] = val
    return entry


def build_manifest(object_cfg: Mapping,
                   variant_entries: Sequence[Mapping]) -> Dict:
    """Assemble the per-object manifest dict.

    Copies present, non-null object metadata (:data:`OBJECT_META_FIELDS`) from
    ``object_cfg``, defaults ``units`` to :data:`DEFAULT_UNITS`, and appends the
    already-built ``variant_entries``. Requires an ``id`` and at least one
    variant; exactly one variant should be flagged ``default`` (see
    :func:`default_variant_index`).
    """
    try:
        oid = object_cfg['id']
    except KeyError:
        raise KeyError("object config requires an 'id'")
    if not variant_entries:
        raise ValueError(f"object {oid!r} has no variant entries")

    manifest: Dict = {'id': oid}
    for field in OBJECT_META_FIELDS:
        val = object_cfg.get(field)
        if val is not None:
            manifest[field] = val
    manifest['units'] = object_cfg.get('units') or DEFAULT_UNITS
    manifest['variants'] = list(variant_entries)

    defaults = [v for v in variant_entries if v.get('default')]
    if len(defaults) != 1:
        raise ValueError(
            f"object {oid!r} must have exactly one default variant, "
            f"got {len(defaults)}")
    return manifest


def index_entry(object_id: str, title: str, manifest_uri: str, *,
                thumb: Optional[str] = None) -> Dict:
    """Build one ``index.json`` ``objects[]`` entry (host archive UI only)."""
    entry: Dict = {'id': object_id, 'title': title, 'manifest': manifest_uri}
    if thumb is not None:
        entry['thumb'] = thumb
    return entry


def build_index(objects: Sequence[Mapping]) -> Dict:
    """Wrap ``index_entry`` results into the optional top-level ``index.json``."""
    return {'objects': list(objects)}


def write_json(data: Mapping, path: PathLike, *, indent: int = 2) -> Path:
    """Write ``data`` as UTF-8 JSON to ``path`` (creating parents); return it."""
    path = Path(path)
    path.parent.mkdir(parents=True, exist_ok=True)
    with path.open('w', encoding='utf8') as f:
        json.dump(data, f, indent=indent, ensure_ascii=False)
        f.write('\n')
    return path

File: viewer/src/manifest.ts
// Types for the per-object manifest the widget consumes.
//
// Mirrors the pipeline's builder in preppy/manifest.py (ADR-0002, amended): one
// manifest per object (one scene) with object-level metadata plus a flat
// `variants[]`, each a self-contained glb referenced by `uri` with a stable `id`
// (the variant suffix / selection + deep-link key). Exactly one variant is flagged
// `default`. The manifest-fetch/validate helpers land in Phase 2 (Task 2.1).

/** One selectable variant: a self-contained glb (meshopt geometry + embedded KTX2). */
export interface Variant {
  /** Stable selection / deep-link key (the pipeline variant `suffix`). */
  id: string;
  /** Human-readable display name; the pipeline falls back to `id`. */
  label: string;
  /** URI of the variant's self-contained glb, relative to the manifest. */
  uri: string;
  /** Present and `true` on exactly one variant. */
  default?: boolean;
  // Optional per-variant provenance overrides (VARIANT_OVERRIDE_FIELDS).
  credit?: string;
  date?: string;
  method?: string;
  description?: string;
}

/** A single object's scene descriptor: metadata + its variants. */
export interface Manifest {
  /** Object id. */
  id: string;
  title?: string;
  /** Optional locale-specific display titles. */
  titles?: Record<string, string>;
  inventory?: string;
  description?: string;
  credit?: string;
  date?: string;
  /** Measurement units for the scene; the pipeline defaults to `"cm"`. */
  units: string;
  variants: Variant[];
}

/** Thrown when a manifest is malformed or a requested variant is missing. */
export class ManifestError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'ManifestError';
  }
}

/**
 * Narrow parsed JSON to a {@link Manifest}, checking the invariants the widget relies
 * on: a non-empty `variants[]` where every entry has a string `id` and `uri`. `units`
 * defaults to `"cm"` and each variant's `label` defaults to its `id` (mirroring the
 * pipeline). Throws {@link ManifestError} on any violation.
 */
export function parseManifest(data: unknown): Manifest {
  if (typeof data !== 'object' || data === null) {
    throw new ManifestError('manifest must be a JSON object');
  }
  const obj = data as Record<string, unknown>;
  if (typeof obj['id'] !== 'string') {
    throw new ManifestError("manifest requires a string 'id'");
  }
  if (!Array.isArray(obj['variants']) || obj['variants'].length === 0) {
    throw new ManifestError('manifest requires a non-empty variants array');
  }

  const variants: Variant[] = obj['variants'].map((raw, i) => {
    if (typeof raw !== 'object' || raw === null) {
      throw new ManifestError(`variant ${i} must be an object`);
    }
    const v = raw as Record<string, unknown>;
    if (typeof v['id'] !== 'string') {
      throw new ManifestError(`variant ${i} requires a string 'id'`);
    }
    if (typeof v['uri'] !== 'string') {
      throw new ManifestError(`variant ${v['id']} requires a string 'uri'`);
    }
    const entry: Variant = {
      id: v['id'],
      label: typeof v['label'] === 'string' ? v['label'] : v['id'],
      uri: v['uri'],
    };
    if (v['default'] === true) {
      entry.default = true;
    }
    for (const field of ['credit', 'date', 'method', 'description'] as const) {
      if (typeof v[field] === 'string') {
        entry[field] = v[field] as string;
      }
    }
    return entry;
  });

  const manifest: Manifest = {
    id: obj['id'],
    units: typeof obj['units'] === 'string' ? obj['units'] : 'cm',
    variants,
  };
  for (const field of ['title', 'inventory', 'description', 'credit', 'date'] as const) {
    if (typeof obj[field] === 'string') {
      manifest[field] = obj[field] as string;
    }
  }
  if (typeof obj['titles'] === 'object' && obj['titles'] !== null) {
    manifest.titles = obj['titles'] as Record<string, string>;
  }
  return manifest;
}

/**
 * Fetch and parse a manifest from `src` (a URL). Returns the manifest and the absolute
 * base URL its variant `uri`s resolve against.
 */
export async function fetchManifest(
  src: string,
  base: string = document.baseURI,
): Promise<{ manifest: Manifest; baseUrl: string }> {
  const url = new URL(src, base).href;
  const response = await fetch(url);
  if (!response.ok) {
    throw new ManifestError(`failed to fetch manifest ${url}: HTTP ${response.status}`);
  }
  const manifest = parseManifest(await response.json());
  return { manifest, baseUrl: url };
}

/**
 * Select the variant to show: the one whose `id` matches `id` if given (throws if
 * absent), else the one flagged `default`, else the first (mirrors the pipeline's
 * default-selection rule).
 */
export function resolveVariant(manifest: Manifest, id?: string): Variant {
  if (id) {
    const found = manifest.variants.find((v) => v.id === id);
    if (!found) {
      throw new ManifestError(`no variant with id '${id}' in manifest '${manifest.id}'`);
    }
    return found;
  }
  return manifest.variants.find((v) => v.default) ?? manifest.variants[0]!;
}

/** Resolve a variant's `uri` against the manifest's base URL to an absolute URL. */
export function resolveVariantUrl(variant: Variant, baseUrl: string): string {
  return new URL(variant.uri, baseUrl).href;
}
